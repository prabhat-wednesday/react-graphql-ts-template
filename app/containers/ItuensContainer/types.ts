export interface Song {
  trackId: number;
  artistName: string;
  artworkUrl100: string;
  collectionName: string;
}

export interface SongData {
  Songs: Song[];
}
