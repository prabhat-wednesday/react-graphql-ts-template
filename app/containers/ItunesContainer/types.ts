import { AnyAction } from '@reduxjs/toolkit';

export interface Song {
  trackId: number;
  artistName: string;
  artworkUrl100: string;
  collectionName: string;
  previewUrl: string;
}

export interface SongData {
  results: Song[];
}

export type SongActionCreator = (payload: string) => AnyAction;

export interface ItuneContainerProps {
  dispatchArtistName: SongActionCreator;
  songData: {
    results: Song[];
  };
  loading: boolean;
  songListError?: string;
}
