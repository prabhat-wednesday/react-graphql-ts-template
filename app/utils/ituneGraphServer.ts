import ApolloClient, { DocumentNode, InMemoryCache } from 'apollo-boost';

export const client = new ApolloClient({
  uri: 'https://itunes.apple.com/search?term=',
  cache: new InMemoryCache()
});

export interface GqlQueryReponse<Data> {
  data?: Data;
  error?: any;
  ok: boolean;
}

export const getQueryResponse = <Data, Variables>(
  query: DocumentNode,
  variables?: Variables
): Promise<GqlQueryReponse<Data>> => {
  console.log(query, 'in graphql server');
  return client
    .query<Data, Variables>({ query, variables })
    .then((res) => {
      console.log(res, 'in graphql');
      if (res.errors) {
        console.log(res);
        return { error: res.errors, ok: false };
      } else {
        console.log(res);
        return { data: res.data, ok: true };
      }
    })
    .catch((err) => {
      return { error: err, ok: false };
    });
};
