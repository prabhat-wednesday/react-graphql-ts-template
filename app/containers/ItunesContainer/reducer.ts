import { prepare } from '@app/utils';
import { createSlice } from '@reduxjs/toolkit';
import { song } from './types';

export interface ituneResponseState {
  loading: boolean;
  songData: {
    resultCount: number;
    results: song[];
  };
  songListError: string;
}

export const initialState: ituneResponseState = {
  loading: false,
  songData: { resultCount: 0, results: [] },
  songListError: ''
};

const ituneSlice = createSlice({
  name: 'Itune',
  initialState,
  reducers: {
    requestGetSongList: {
      reducer: (state) => {
        state.loading = true;
      },
      prepare
    },
    successGetSongList: (state, action) => {
      state.songData = action.payload;
      state.songListError = '';
      state.loading = false;
    },
    failureGetSongList: (state, action) => {
      state.songData.resultCount = 0;
      state.songData.results = [];
      state.loading = false;
      state.songListError = action.payload;
    }
  }
});

export const { requestGetSongList, successGetSongList, failureGetSongList } = ituneSlice.actions;

export default ituneSlice.reducer;
