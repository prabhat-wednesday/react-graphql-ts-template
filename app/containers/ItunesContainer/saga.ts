import { takeLatest, call, put } from 'redux-saga/effects';
import { requestGetSongList, successGetSongList, failureGetSongList } from '../SongProviderContainer/reducer';
import { getItune } from '@services/apiUtils';

export function* fetchDataFromItune(action: any): Generator<any, any> {
  const { artistName, pageNumber, pageSize } = action.payload ?? { artistName: '', pageNumber: 0, pageSize: 0 };
  const res: any = yield call(getItune, { artistName, pageNumber, pageSize });
  const { data, ok, error } = res;
  if (ok) {
    yield put(successGetSongList(data));
  } else {
    yield put(failureGetSongList(error));
  }
}

function* ituneCallSaga() {
  yield takeLatest(requestGetSongList.toString(), fetchDataFromItune);
}

export default ituneCallSaga;
