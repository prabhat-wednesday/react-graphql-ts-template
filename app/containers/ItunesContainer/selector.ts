import { createSelector } from 'reselect';
import get from 'lodash-es/get';
import { initialState } from './reducer';

const selectItuneState = (state: any) => state.ituneReducer || initialState;

export const selectLoading = () => createSelector(selectItuneState, (substate) => get(substate, 'loading'));
export const selectSongData = () => createSelector(selectItuneState, (substate) => get(substate, 'songData'));
export const selectSongListError = () => createSelector(selectItuneState, (substate) => get(substate, 'error'));
