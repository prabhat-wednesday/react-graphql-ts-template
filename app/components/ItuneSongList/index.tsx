import React from 'react';
import { Row } from 'antd';
import { get } from 'lodash-es';
import styled from 'styled-components';
import { Song } from '@app/containers/ItunesContainer/types';
import { colors, media } from '@app/themes';
import { For } from '../For';
import ItuneCard from '../ItuneCard';

interface ItuneSongListProps {
  songData: {
    results?: Song[];
  };
}

const CustomRow = styled(Row)`
  && {
    background-color: ${colors.cardContainerBg};
    gap: 2.5rem;

    ${media.lessThan('tablet')`
    justify-content: center;`}
  }
`;

const ItuneSongList = ({ songData }: ItuneSongListProps) => {
  const results = get(songData, 'results', []);
  return (
    <>
      <For ParentComponent={CustomRow} renderItem={(song: Song) => <ItuneCard {...song} />} of={results} />
    </>
  );
};

export default ItuneSongList;
