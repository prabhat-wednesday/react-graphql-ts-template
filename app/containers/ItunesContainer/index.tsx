import React, { ChangeEvent } from 'react';
import debounce from 'lodash-es/debounce';
import { createStructuredSelector } from 'reselect';
import { connect } from 'react-redux';
import { injectSaga } from 'redux-injectors';
import { compose } from '@reduxjs/toolkit';
import { selectSongListError, selectLoading, selectSongData } from './selector';
import { requestGetSongList } from './reducer';
import ituneCallSaga from './saga';
import { ItuneContainerProps } from './types';
import ItuneSongList from '@app/components/ItuneSongList';
import styled from 'styled-components';
import { Input } from 'antd';
import { media } from '@app/themes';
import { T } from '@app/components';

const InputContainer = styled.div`
  && {
    display: flex;
    flex-direction: column;
    align-items: center;
    margin: 0.5rem auto;
  }
`;

// const CustomLabel = styled.label`
//   && {
//     display: block;
//     font-size: 1.5rem;
//   }
// `;

const CustomInput = styled(Input)`
  && {
    width: 30%;
    display: block;

    ${media.lessThan('tablet')`
    width: 80%;`}
  }
`;

const ItunesContainer = ({ dispatchArtistName, songData }: ItuneContainerProps) => {
  const handleOnChange = debounce((e: ChangeEvent<HTMLInputElement>) => {
    const artistSearch = e.target.value;
    if (artistSearch.trim()) {
      dispatchArtistName(artistSearch);
    }
  }, 500);

  return (
    <div>
      <InputContainer>
        <T id="song_search_default" />
        <CustomInput onChange={(e) => handleOnChange(e)} type="text" />
      </InputContainer>
      <ItuneSongList songData={songData} />
    </div>
  );
};

const mapStateToProps = createStructuredSelector({
  loading: selectLoading(),
  songData: selectSongData(),
  error: selectSongListError()
});

export function mapDispatchToProps(dispatch: (arg0: { type: string }) => void) {
  return {
    dispatchArtistName: (payload: string) => dispatch(requestGetSongList(payload))
  };
}

const withConnect = connect(mapStateToProps, mapDispatchToProps);

export default compose(withConnect, injectSaga({ key: 'ituneComponent', saga: ituneCallSaga }))(ItunesContainer);
