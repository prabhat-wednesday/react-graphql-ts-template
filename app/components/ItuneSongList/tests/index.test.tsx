import React from 'react';
import { renderProvider } from '@app/utils/testUtils';
import ItuneSongList from '..';
import { SongData } from '@app/containers/ItunesContainer/types';

describe('<ItuneSongList/> ', () => {
  const songData: SongData = {
    results: [
      {
        trackId: 1,
        artistName: 'Yung Xiety',
        artworkUrl100:
          'https://is1-ssl.mzstatic.com/image/thumb/Music112/v4/3b/cc/31/3bcc315a-e58f-c678-8d05-b7d62346070f/1d724fe3-daaa-4825-91f0-a47e70a438bd.jpg/100x100bb.jpg',
        collectionName: 'Arijit Singh (Mashup) - Single'
      }
    ]
  };

  it('should render and match the snapshot', () => {
    const { baseElement } = renderProvider(<ItuneSongList songData={songData} />);
    expect(baseElement).toMatchSnapshot();
  });
});
