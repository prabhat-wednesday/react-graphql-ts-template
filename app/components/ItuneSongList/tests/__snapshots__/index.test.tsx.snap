// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ItuneSongList/>  should render and match the snapshot 1`] = `
<body>
  <div>
    <div
      class="ant-row ItuneSongList__CustomRow-sc-1kmczkg-0 ljogwX"
      data-testid="for"
      orientation="row"
    >
      <div
        class="ant-card ant-card-bordered ant-card-hoverable ItuneCard__CustomCard-sc-1g3ywlw-1 kNwYqr"
      >
        <div
          class="ant-card-cover"
        >
          <img
            class="ItuneCard__StyledImg-sc-1g3ywlw-2 iLWLsw"
            loading="lazy"
            src="https://is1-ssl.mzstatic.com/image/thumb/Music112/v4/3b/cc/31/3bcc315a-e58f-c678-8d05-b7d62346070f/1d724fe3-daaa-4825-91f0-a47e70a438bd.jpg/100x100bb.jpg"
          />
        </div>
        <div
          class="ant-card-body"
        >
          <h1
            class="ItuneCard__H1-sc-1g3ywlw-0 dqixeH"
          >
            Card Title
            : 
            Yung Xiety
          </h1>
          <div
            class="ant-card-meta"
          >
            <div
              class="ant-card-meta-detail"
            >
              <div
                class="ant-card-meta-title"
              >
                Arijit Singh (Mashup) - Single
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
